### DISCUSSION TITLE
igl::eigs returns different eigen vectors with the same input.

### DISCUSSION THREAD

**Discussion by lukkio88** [libigl/libigl#2206] [Q&A] [answered]
Hello. I have a wrapper function that is called with the same mesh (so essentially call the function twice in a row with the same input).
The function is as simple as

```
void generateEigenFunctions(const Eigen::MatrixXd& vertices, const Eigen::MatrixXi& faces, Eigen::MatrixXd& eigenVectors, Eigen::VectorXd& eigenValues, int numberOfEigenFunctions) {
    Eigen::SparseMatrix<double> laplacian, massMatrix;
    igl::cotmatrix(vertices, faces, laplacian);
    igl::massmatrix(vertices, faces, igl::MASSMATRIX_TYPE_DEFAULT, massMatrix);
    igl::eigs(laplacian, massMatrix, numberOfEigenFunctions, igl::EIGS_TYPE_SM, eigenVectors, eigenValues);
}
```

Which is essentially the code you have: https://libigl.github.io/tutorial/#eigen-decomposition

However, despite the same input, I do get always the last eigenValue extracted different:

In this example see the smallest:

1.48503e-13    -68.6417    -92.5271    -148.268    -167.977     -218.91 //after first call
1.27533e-13    -68.6417    -92.5271    -148.268    -167.977     -218.91 //after second

Is there a reason for this?

**Comment by alecjacobson** [comment, marked as answer]
yes, this is somewhat expected given the implementation uses a random initial guess `Eigen::…:Random`.

There are some strong opinions about the best way to use randomization in practice. But in the meantime you can control the Eigen random seed https://stackoverflow.com/a/67387606/148668

**Reply by lukkio88** [reply]
Yes, I realized this right after. I wonder if Using Spectra might avoid this shortcoming (together with the fact that also for some meshes I cannot get more than 5 eigen functions as well).

**Reply by alecjacobson** [reply]
Not sure about the random start but spectra should be much more reliable and efficient for more than one or two eigen modes. There are some tricks to getting it in the right mode for low frequencies.
